From 9d5b3cd1149b9232e17c70923f6831fe12f58f34 Mon Sep 17 00:00:00 2001
From: Peter Harliman Liem <peter.harliman.liem@intel.com>
Date: Mon, 25 Mar 2019 09:27:15 +0800
Subject: [PATCH] DRVLIB_SW-2236 - Update gphy RCAL info retrieval for B0 Step

B0 step requires a different method with A1 for RCAL info retrieval:
- B0 --> RCAL info read from CDB status register
- A1 --> RCAL info read from Fused register

GPHY driver code is updated for this reason
---
 drivers/net/ethernet/lantiq/xrx500_phy_fw.c | 16 ++++++++++++----
 1 file changed, 12 insertions(+), 4 deletions(-)

diff --git a/drivers/net/ethernet/lantiq/xrx500_phy_fw.c b/drivers/net/ethernet/lantiq/xrx500_phy_fw.c
index b1ac073beaee..49a238a1dd94 100644
--- a/drivers/net/ethernet/lantiq/xrx500_phy_fw.c
+++ b/drivers/net/ethernet/lantiq/xrx500_phy_fw.c
@@ -15,6 +15,7 @@
 #include <linux/reset.h>
 #include <linux/regmap.h>
 #include <linux/mfd/syscon.h>
+#include <lantiq.h>
 
 #define XRX500_GPHY_NUM 5 /* phy2-5 + phyf */
 struct xrx500_reset_control {
@@ -86,6 +87,7 @@ static u32 xrx500_gphy[] = {
 /* GPHY CDB FCSI */
 #define PRX300_GPHY_CDB_FCSI_PLL_CFG1 0x0
 #define PRX300_GPHY_CDB_FCSI_PLL_CFG2 0x4
+#define PRX300_GPHY_CDB_FCSI_PLL_RCALSTAT 0x20
 #define PRX300_GPHY_CDB_FCSI_PLL_RCMSTAT 0x28
 #define PRX300_GPHY_CDB_FCSI_PLL_RCMCFG 0x2c
 /* GPHY CDB PDI */
@@ -195,10 +197,16 @@ static u32 prx300_gphy_config_rcal_rcm(struct xway_gphy_data *priv)
 	u32 val;
 	int retry;
 
-	/* get rcal from fused register (upper 4-bits) */
-	regmap_read(priv->chipid_syscfg, PRX300_FUSE_REDUND_1,
-		    &val);
-	val = val >> 28;
+	if (ltq_get_soc_rev() == 1) {
+		/* B-Step: get rcal from cdb register */
+		val = gsw_reg_r32(priv->fcsi_base,
+				  PRX300_GPHY_CDB_FCSI_PLL_RCALSTAT) & 0xF;
+	} else {
+		/* A-Step: get rcal from fused register (upper 4-bits) */
+		regmap_read(priv->chipid_syscfg, PRX300_FUSE_REDUND_1,
+			    &val);
+		val = val >> 28;
+	}
 
 	/* no fused values, simply use default settings */
 	if (!val)
